### ISSUE TITLE
jsondecode() does not work with surrogate pairs

### ISSUE THREAD

**Issue by ZyX-I** [vim/vim#613]
Consider the following code:

```
vim -u NONE -i NONE -c 'echo jsondecode("\"\\uD800\\uDC00\"")'
```

This will echo `<d800><dc00>`, but it actually should echo `𐀀` (U+10000) because JSON encodes code points above U+FFFF using surrogate pairs.

Note that `"\uD800"` or `"\uDC00"` alone cannot be turned into a valid UTF-8, but specification still allows such strings (result of parsing is implementation-defined, up to fatal runtime error):

>    However, the ABNF in this specification allows member names and
>    string values to contain bit sequences that cannot encode Unicode
>    characters; for example, "\uDEAD" (a single unpaired UTF-16
>    surrogate).  Instances of this have been observed, for example, when
>    a library truncates a UTF-16 string without checking whether the
>    truncation split a surrogate pair.  The behavior of software that
>    receives JSON texts containing such values is unpredictable; for
>    example, implementations might return different values for the length
>    of a string value or even suffer fatal runtime exceptions.

**Comment by brammool** [comment]
Nikolai Pavlov wrote:

> Consider the following code:
> 
> ```
> vim -u NONE -i NONE -c 'echo jsondecode("\"\\uD800\\uDC00\"")'
> ```
> 
> This will echo `<d800><dc00>`, but it actually should echo `𐐀` (U+10000) because JSON encodes code points above U+FFFF using surrogate pairs.
> 
> Note that `"\uD800"` or `"\uDC00"` alone cannot be turned into a valid UTF-8, but specification still allows such strings (result of parsing is implementation-defined, up to fatal runtime error):
> 
> >    However, the ABNF in this specification allows member names and
> >    string values to contain bit sequences that cannot encode Unicode
> >    characters; for example, "\uDEAD" (a single unpaired UTF-16
> >    surrogate).  Instances of this have been observed, for example, when
> >    a library truncates a UTF-16 string without checking whether the
> >    truncation split a surrogate pair.  The behavior of software that
> >    receives JSON texts containing such values is unpredictable; for
> >    example, implementations might return different values for the length
> >    of a string value or even suffer fatal runtime exceptions.

I didn't implement this UTF-16 encoding yet.  It was a bad choice to use
this encoding...  Nevertheless, it's part of the standard.

## 

From "know your smileys":
 ~#:-(  I just washed my hair, and I can't do nuthin' with it.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by Milly** [comment]
This issue is fixed at v7.4.1434 (b6ff811).

**Comment by k-takata** [comment]
Closing then.
